gh-95299: Stop installing setuptools as a part of ensurepip and venv

pradyunsg opened the pull request:
This PR removes the bundled setuptools wheel from ensurepip, and stops installing setuptools in environments created by venv.

---

I based this PR off of my understanding of venv and ensurepip, and using the following search to locate all the relevant bits of code + ensuring that tests are happy locally.

![Screenshot 2023-01-14 at 17 20 01](https://user-images.githubusercontent.com/3275593/212486567-e396f2b1-63a5-46d4-93e4-c4405fd731b7.png)

In this PR, I've intentionally not touched the example in https://docs.python.org/3.12/library/venv.html#an-example-of-extending-envbuilder. That example is largely outdated and is likely better served being updated in a dedicated PR for it.

<!-- gh-issue-number: gh-95299 -->
* Issue: gh-95299
<!-- /gh-issue-number -->

pradyunsg committed:
Stop installing setuptools as a part of ensurepip and venv

This is no longer necessary since newer versions of pip are able to
operate in environments that do not contain setuptools.

pradyunsg committed:
Use environment variables to pass data into test setup script

This avoids meddling with `sys.argv` and removes reliance on careful
control over the arguments passed to the script.

pradyunsg committed:
Update test_cppext to stop invoking setuptools-based script as a CLI

This allows the "package" to be built using regular Python packaging
workflows instead of relying on `setuptools` being present in the
virtual environment.

hugovk commented on .github/workflows/verify-ensurepip-wheels.yml:
```diff
@@ -28,5 +28,5 @@ jobs:
         with:
           python-version: '3'
-      - name: Compare checksums of bundled pip and setuptools to ones published on PyPI
+      - name: Compare checksums of bundled pip to ones published on PyPI
```
Now singular 
```suggestion
      - name: Compare checksum of bundled pip to the one published on PyPI
```

pradyunsg committed:
Correct the name of step in GitHub Action

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

vsajip reviewed (approved):
The changes seem reasonable to me, though I'm only the code owner for the venv and test_venv bits.

pfmoore commented on Lib/test/test_cppext.py:
```diff
@@ -12,7 +16,7 @@
 
 
-SETUP_TESTCPPEXT = support.findfile('setup_testcppext.py')
+PKG_CPPEXTDATA = os.path.dirname(support.findfile('setup.py', subdir="cppextdata"))
```
This is the unadorned directory name, which is treated below by pip as a *package* name, not a directory.

Add an `os.path.abspath()` round this to disambiguate?

merwok commented on Lib/ensurepip/__init__.py:
```diff
@@ -153,10 +151,10 @@ def _bootstrap(*, root=None, upgrade=False, user=False,
     # following scripts (X.Y == running Python version):
     #
-    #   pip, pipX, pipX.Y, easy_install, easy_install-X.Y
+    #   pip, pipX, pipX.Y
```
more mentions of easy_install on lines 161 and 164

merwok commented on Lib/test/test_cppext.py:
```diff
@@ -59,29 +65,26 @@ def _check_build(self, std_cpp03, extension_name):
                                       stderr=subprocess.STDOUT,
-                                      text=True)
+                                      text=True,
+                                      env=env)
```
I would’ve put `env` right after `cmd`, as they are related to the process model, whereas `text` is python processing.
Very minor of course.

pradyunsg committed:
Drop most remaining references to `easy_install`

Since setuptools will no longer get installed, this will no longer be
provided out-of-the-box.

pradyunsg committed:
fixup! Update test_cppext to stop invoking setuptools-based script as a CLI

pradyunsg committed:
fixup! Update test_cppext to stop invoking setuptools-based script as a CLI

pradyunsg committed:
Add a "What's New" entry for the setuptools removal

This provides context, keywords and some guidance on what this change
would do.

pradyunsg committed:
fixup! Add a "What's New" entry for the setuptools removal

pradyunsg committed:
fixup! Add a "What's New" entry for the setuptools removal

CAM-Gerlach commented on Doc/whatsnew/3.12.rst:
```diff
@@ -541,6 +541,23 @@ Removed
   project can be installed: it still provides ``distutils``.
   (Contributed by Victor Stinner in :gh:`92584`.)
 
+* Remove the bundled setuptools wheel from ``ensurepip``, and stop installing setuptools in environments created by ``venv``.
```
```suggestion
* Remove the bundled setuptools wheel from :mod:`ensurepip`,
  and stop installing setuptools in environments created by :mod:`venv`.
```

* Cross-reference the respective modules
* Line wrap

pradyunsg committed:
Update Doc/whatsnew/3.12.rst

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

arhadthedev committed:
Merge branch 'main' into remove-setuptools-ensurepip

pradyunsg committed:
Merge branch 'main' into remove-setuptools-ensurepip

pradyunsg committed:
Update "Creating virtual environments" section to drop reference

merwok commented on Doc/using/venv-create.inc:
```diff
@@ -61,12 +61,16 @@ The command, if run with ``-h``, will show the available options::
       --prompt PROMPT       Provides an alternative prompt prefix for this
                             environment.
-      --upgrade-deps        Upgrade core dependencies: pip setuptools to the
+      --upgrade-deps        Upgrade core dependencies: pip to the
```
```suggestion
      --upgrade-deps        Upgrade core dependencies (pip) to the
```

pradyunsg replied:
Done, and updated the code that this is supposed to be equivalent to as well.

pradyunsg committed:
Use the style suggested in review

pradyunsg committed:
Use dirname on the absolute path

pradyunsg committed:
Locate cppextdata with plain filesystem lookups

Doing anything more elaborate was flaky on certain buildbots, and
hopefully this is good enough.

merwok commented on .github/workflows/verify-ensurepip-wheels.yml:
```diff
@@ -1,4 +1,4 @@
-name: Verify bundled pip and setuptools
+name: Verify bundled pip
```
For this file I think it would be fine to not be specific in the names, like:
Verify bundled wheels
Compare checksums of bundled wheels to the ones on PyPI

CAM-Gerlach commented on Lib/test/test_cppext.py:
```diff
@@ -72,16 +81,9 @@ def run_cmd(operation, cmd):
+        # Build and install the C++ extension
         cmd = [python, '-X', 'dev',
-               SETUP_TESTCPPEXT, 'install']
+               '-m', 'pip', 'install', PKG_CPPEXTDATA]
```
LGTM on the docs now, thanks.

You you're probably already on it, but seems like this is failing on [some of the buildbots](https://buildbot.python.org/all/#/builders/465/builds/1049) with

```
subprocess.CalledProcessError: Command '['env/bin/python', '-X', 'dev', '-m', 'pip', 'install', '/buildbot/buildarea/cpython/pull_request.ware-gentoo-x86.installed/build/target/lib/python3.12/test/cppextdata']' returned non-zero exit status 1.
```

And reporting the following error:

```
ERROR: Invalid requirement: '/buildbot/buildarea/cpython/pull_request.ware-gentoo-x86.installed/build/target/lib/python3.12/test/cppextdata'
```

pradyunsg replied:
I'm aware -- these specific buildbots have been consistently failing since the start, hence me triggering runs on them with smaller changes to the test setup. :)

I've not been able to reproduce the error so far outside of those environments, and not had the time/knowledge to investigate what's going on with these buildbots.

CAM-Gerlach replied:
Okay, thanks. I've dug in a bit and found the following (which you probably already did, but just documenting this for completeness for any others reading this):

* s390x RHEL7 & 8 refleaks failures with `test_asyncio` appears to be pre-existing pre-existing and error identically on other older and newer PRs
* s390x Fedora LTO + PGO failure with missing `make` target likewise fails the same way with older and newer PRs, so evidently also unrelated
* AMD64 Windows11 Bigmem test error with `test_hashlib` does not appear to be exactly consistent with others, some of which passed and some failed with different errors, but it has very little build history to go on, it appears rather flaky (it is bigmem after all) and the test failure here is only on a single run and doesn't seem obviously related, so maybe we can defer it for now.
* The remaining failures all occur if and only if the buildbot is an `Installed` buildbot (all Linux, various distros and platforms), where `pip` apparently cannot find the `cppextdata` directory with the C extension to build at the installed path. Given the search path, `__file__ / "cppextdata"` is different than for the non-installed versions (under `build/target/lib/python3.12/test/cppextdata` rather than `build/Lib/test/cppextdata`, I'm assuming the `cppextdata` is not present in that specific directory when installed, and is instead somewhere else or not installed/copied.

As to where specifically it is installed/copied I'm not sure and I'm not expert on FHS/Linux filesystem layouts, but I would think that's something that can be reproduced after installing in on a Linux box, VM a similar manner (or maybe just asking Filipe who is the expert at Linux install schemes, or perhaps its something that one of the other Linux gurus would know)?

vstinner commented on Lib/test/test_cppext.py:
```diff
@@ -31,6 +34,8 @@ def test_build_cpp03(self):
         '-fsanitize' in (sysconfig.get_config_var('PY_CFLAGS') or ''),
         'test does not work with analyzing builds')
+    # the test uses pip which needs a TLS connection to PyPI
+    @unittest.skipIf(ssl is None, 'No ssl module')
```
It would be better to have a decorator like ``@test.support.requires_venv_with_pip``. requires_venv_with_pip() checks also for zlib and ctypes, so it's more complicated than just ssl in practice.

pradyunsg replied:
We already have that decorator on this test.

pradyunsg committed:
Merge branch 'main' into remove-setuptools-ensurepip

pradyunsg committed:
bundled pip -> bundled wheels

pradyunsg committed:
Unbreak `test_cppext`

pradyunsg committed:
:art:
